### PR TITLE
Add 'Gateway Bouncer' test scenario to the 'perf' suite

### PR DESCRIPTION
**Pull request by philrud**
This scenario spins up a minimal 'isolated' setup that can be used for testing
or the IngressGateway in ways that are either destructive or require setup
that is not compatible with the default gateways under `istio-system`.

One the setup is up and running the following capabilities are exercised:

1. External traffic is simulated by Fortio and will go via external load
balancer through the IngressGateway (a new LoadBalancer will be created
automatically by the install script and its external IP address will be used).
1. A rolling restart of the IngressGateway will be periodically triggered while
the corresponding Pilot is down (and eventually brought up), so the readiness
checks of the IngressGateway can be verified. IngressGateway should not become
ready until the Pilot is back up and a configuration has been received.
1. Rolling restart of the gateway should be transparent for the clients. Fortio
Client is configured to crash upon connectivity errors. 'Connection refused'
will result in a restart of Fortio Client pods. Prometheus metrics on pod
status can be used to verify whether rolling redeploy of the IngressGateway
results in Fortio Client restarts.
1. The IngressGateway routes are configured using standard K8S Ingress with a
custom Ingress class label to make sure multiple Ingress configurations can
co-exist in a single cluster.

Use the following command to (re)install the scenario into a new/existing
namespace (set `DRY_RUN=1` for a dry run):
```
NAMESPACE=gateway-bouncer ./setup.sh
```

### CODE REVIEW

**Comment by philrud** [comment]
@mandarjog @costinm could you help with the review please? Also if you want to give it a spin just use a custom namespace and delete it when you're done with it. It should be pretty compact and it won't pollute istio-system or pull any extra things into existing cluster in general. If this is looking good for you I'll make another PR for SDS+certmanager using the same exact approach.

File: perf/istio/gateway-bouncer/setup.sh [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+  helm -n $NAMESPACE template \
+    --set namespace=$NAMESPACE \
+    . > "${OUTFILE}"
+
+  if [[ -z "${DRY_RUN}" ]]; then
+    kubectl -n $NAMESPACE apply -f "${OUTFILE}"
+
+    # Waiting until LoadBalancer is created and retrieving the assigned
+    # external IP address.
+    while : ; do
```
Comment (howardjohn) [inline perf/istio/gateway-bouncer/setup.sh]: One concern here is if a LB isn't being created, will this just spin forever with no indication?

Or does the `set -x` show what is being done?

File: perf/istio/gateway-bouncer/setup.sh [hunk 2]
```diff
@@ -0,0 +1,46 @@
+#/bin/bash
+set -ex
+
+NAMESPACE=${NAMESPACE:?"New/existing NAMESPACE to install this scenario into"}
```
Comment (howardjohn) [inline perf/istio/gateway-bouncer/setup.sh]: Most (maybe all) other tests set a default namespace. We should probably clean up the various scenarios to behave the same way - probably to have namespace configurable or have a default value?

File: perf/istio/gateway-bouncer/templates/fortio-server.yaml [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+        image: {{ .Values.fortioImage }}
+        imagePullPolicy: Always
+        ports:
+         - containerPort: 8080
+         - containerPort: 8079
+        args:
+          - server
+        resources:
+          requests:
+            cpu: 1000m
```
Comment (howardjohn) [inline perf/istio/gateway-bouncer/templates/fortio-server.yaml]: These seem a bit high, does fortio need this much?

File: perf/istio/gateway-bouncer/setup.sh (hunk 1 above)
Comment (philrud) [reply perf/istio/gateway-bouncer/setup.sh]: Yes, -x provides quite a bit of feedback, try running it -- pretty self explanatory. Hopefully in the future we can streamline the installation process, so we won't have to rely on a bunch of ad-hoc scripts.

File: perf/istio/gateway-bouncer/setup.sh (hunk 2 above)
Comment (philrud) [reply perf/istio/gateway-bouncer/setup.sh]: Yes, that's the idea, to be able to easily spin up this kind of test in isolation (where it makes sense). Namespace is currently a required param (similar to how DOMAIN is required in 'allconfig') as we definitely don't want to install this particular scenario into 'istio-system'.

File: perf/istio/gateway-bouncer/templates/fortio-server.yaml (hunk 3 above)
Comment (philrud) [reply perf/istio/gateway-bouncer/templates/fortio-server.yaml]: It's convenient to have at least one core, so I can easily dial in more load by increasing the number of client replicas if we're interested in what is happening under the load. I only went with one core though (allconfig setup is using like 4x cores for the Fortio Server -- I don't think we need that much for this particular scenario though).

File: perf/istio/gateway-bouncer/setup.sh (hunk 2 above)
Comment (costinm) [reply perf/istio/gateway-bouncer/setup.sh]: How about picking a default name, so someone running this doesn't have to worry about ? 
'gateway-bouncer' is as good name as any.

File: perf/istio/gateway-bouncer/setup.sh [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+  local DIRNAME="${1:?"output dir"}"
+  local OUTFILE="${DIRNAME}/gateway_bouncer.yaml"
+  local INGRESS_IP=""
+
+  helm -n $NAMESPACE template \
+    --set namespace=$NAMESPACE \
+    . > "${OUTFILE}"
+
+  if [[ -z "${DRY_RUN}" ]]; then
+    kubectl -n $NAMESPACE apply -f "${OUTFILE}"
```
Comment (costinm) [inline perf/istio/gateway-bouncer/setup.sh]: Happy to see use of helm template + apply - but for this you need to use the --prune option and make sure you have a release label ( see the istio-install env.sh for an example )

File: perf/istio/gateway-bouncer/setup.sh [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+      fi
+    done
+
+    # Populating a ConfigMap with the external IP address and restarting the
+    # client to pick up the new version of the ConfigMap.
+    kubectl -n $NAMESPACE delete configmap fortio-client-config
+    kubectl -n $NAMESPACE create configmap fortio-client-config \
+      --from-literal=external_addr=$INGRESS_IP
+    kubectl -n $NAMESPACE patch deployment istio-ingress-$NAMESPACE \
+      -p "{\"spec\":{\"template\":{\"metadata\":{\"labels\":{\"date\":\"`date +'%s'`\"}}}}}"
```
Comment (costinm) [inline perf/istio/gateway-bouncer/setup.sh]: Generally I prefer patching an annotation, not label ( labels create secondary indexes and trigger a lot of other side effects )

File: perf/istio/gateway-bouncer/templates/fortio-server.yaml (hunk 3 above)
Comment (costinm) [reply perf/istio/gateway-bouncer/templates/fortio-server.yaml]: We can probably keep it at 0.2 for 200 qps.  Usually it's 3000 qps per core for sidecar, fortio can handle much more.

**Pull request merged**